openmls/src/treesync/treekem.rs

```diff
@@ -0,0 +1,314 @@
+//! Module to encrypt and decrypt update paths for a [`TreeSyncDiff`] instance.
+//!
+//! # About
+//!
+//! This module contains structs and functions to encrypt and decrypt path
+//! updates for a [`TreeSyncDiff`] instance.
+use std::collections::HashSet;
+
+use tls_codec::{Error as TlsCodecError, TlsDeserialize, TlsSerialize, TlsSize, TlsVecU32};
+
+use openmls_traits::{crypto::OpenMlsCrypto, types::HpkeCiphertext, OpenMlsCryptoProvider};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    binary_tree::LeafIndex,
+    ciphersuite::{Ciphersuite, HpkePublicKey},
+    config::ProtocolVersion,
+    key_packages::{KeyPackage, KeyPackageError},
+    messages::{
+        proposals::AddProposal, EncryptedGroupSecrets, GroupSecrets, PathSecret, PathSecretError,
+    },
+    schedule::{CommitSecret, JoinerSecret, PreSharedKeys},
+};
+
+use super::{
+    diff::TreeSyncDiff,
+    node::parent_node::{ParentNode, ParentNodeError, PlainUpdatePathNode},
+    TreeSyncDiffError, TreeSyncError,
+};
+
+impl<'a> TreeSyncDiff<'a> {
+    /// Encrypt the given `path` to the nodes in the copath resolution of the
+    /// owner of this [`TreeSyncDiff`]. The `group_context` is used in the
+    /// encryption of the nodes, while the `exclusion_list` is used to filter
+    /// target leaves from the encryption targets. The given [`KeyPackage`] is
+    /// included in the resulting [`UpdatePath`].
+    ///
+    /// Returns the encrypted path (i.e. an [`UpdatePath`] instance).
+    pub(crate) fn encrypt_path(
+        &self,
+        backend: &impl OpenMlsCryptoProvider,
+        ciphersuite: &Ciphersuite,
+        path: &[PlainUpdatePathNode],
+        group_context: &[u8],
+        exclusion_list: &HashSet<&LeafIndex>,
+        key_package: &KeyPackage,
+    ) -> Result<UpdatePath, TreeKemError> {
+        let copath_resolutions = self.copath_resolutions(self.own_leaf_index(), exclusion_list)?;
+
+        let mut update_path_nodes = Vec::new();
+        // Encrypt the secrets
+        for (node, resolution) in path.iter().zip(copath_resolutions.iter()) {
```

Comment:
Do we know the lengths of the two match?
And you can pre-allocate the vector again.